File: app/src/main/java/ru/tpu/russiantpu/main/fragments/profile/DocumentsFragment.java
package ru.tpu.russiantpu.main.fragments.profile;

import android.app.Activity;
import android.content.pm.PackageManager;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.core.widget.ContentLoadingProgressBar;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentManager;
import androidx.recyclerview.widget.RecyclerView;

import java.util.ArrayList;

import ru.tpu.russiantpu.R;
import ru.tpu.russiantpu.dto.DocumentDTO;
import ru.tpu.russiantpu.main.dataAdapters.ClickListener;
import ru.tpu.russiantpu.main.dataAdapters.DocumentsDataAdapter;
import ru.tpu.russiantpu.utility.DownloadFileFromUrl;
import ru.tpu.russiantpu.utility.SharedPreferencesService;
import ru.tpu.russiantpu.utility.StartActivityService;
import ru.tpu.russiantpu.utility.ToastService;
import ru.tpu.russiantpu.utility.callbacks.GenericCallback;
import ru.tpu.russiantpu.utility.requests.GsonService;
import ru.tpu.russiantpu.utility.requests.RequestService;

public class DocumentsFragment extends Fragment {

    private RequestService requestService;
    private ContentLoadingProgressBar progressBar;

    //переменные файла
    private String url; //ссылка на скачивание
    private String fileName; //название файла

    private String token; //токен для получения файла

    @Nullable
    @Override
    public View onCreateView(@NonNull final LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        final Activity activity = getActivity();
        final FragmentManager fragmentManager = getFragmentManager();

        final View layoutInflater = inflater.inflate(R.layout.fragment_documents, container, false);
        final RecyclerView recyclerView = layoutInflater.findViewById(R.id.doc_list); //список
        progressBar = layoutInflater.findViewById(R.id.progress_bar);

        //вспомогательные классы
        final SharedPreferencesService sharedPreferencesService = new SharedPreferencesService(activity);
        final GsonService gsonService = new GsonService();
        final ToastService toastService = new ToastService(getContext());
        requestService = new RequestService(sharedPreferencesService, new StartActivityService(activity));

        final ArrayList<DocumentDTO> items = new ArrayList<>();
        final DocumentsDataAdapter adapter = new DocumentsDataAdapter(getContext(), items);

        //получение информации о юзере из sharedPreferences
        final String email = sharedPreferencesService.getEmail();
        final String language = sharedPreferencesService.getLanguageId();
        token = sharedPreferencesService.getToken();

        adapter.setOnItemClickListener(new ClickListener() {
            @Override
            public void onItemClick(int position, View v) {
                url = items.get(position).getUrl();
                fileName = items.get(position).getFileName();
                DownloadFileFromUrl.downloadFile(url, fileName, token, DocumentsFragment.this);
            }
            //пока не используется, оставлен на будущее
            @Override
            public void onItemLongClick(int position, View v) {
            }
        });
        recyclerView.setAdapter(adapter);

        progressBar.show(); //включаем прогресс бар

        //реализация коллбека - что произойдет при получении данных с сервера
        GenericCallback<String> callback = new GenericCallback<String>() {
            @Override
            public void onResponse(String jsonBody) {
                items.addAll(gsonService.fromJsonToArrayList(jsonBody, DocumentDTO.class));

                //отрисовываем список статей в потоке интерфейса
                activity.runOnUiThread(new Runnable() {
                    @Override
                    public void run() {
                        adapter.notifyDataSetChanged();
                        progressBar.hide(); //выключаем прогресс бар
                    }

                });
            }

            @Override
            public void onError(String message) {
                //выключаем прогресс бар
                activity.runOnUiThread(new Runnable() {
                    @Override
                    public void run() {
                        progressBar.hide();
                    }
                });
                toastService.showToast(message);
            }

            @Override
            public void onFailure(String message) {
                //выключаем прогресс бар
                activity.runOnUiThread(new Runnable() {
                    @Override
                    public void run() {
                        progressBar.hide();
                    }
                });
                toastService.showToast(R.string.docs_get_error);
            }
        };

        //запрос за получение документов для пользователя (по email)
        requestService.doRequest("document", callback, token, language, "email", email);

        return layoutInflater;
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        //если юзер дал разрешение на сохранение файла через диалоговое окно, пытаемся скачать файл повторно
        if (grantResults.length > 0 && grantResults[0] == PackageManager.PERMISSION_GRANTED){
            DownloadFileFromUrl.downloadFile(url, fileName, token, DocumentsFragment.this);
        }
    }

    @Override
    public void onDetach() {
        super.onDetach();
        //при закрытии фрагмента отменяем все запросы
        requestService.cancelAllRequests();
        //выключаем прогрессбар
        progressBar.hide();
    }
}

File: app/src/main/java/ru/tpu/russiantpu/main/fragments/profile/DocumentsFragment.kt
package ru.tpu.russiantpu.main.fragments.profile

import android.content.pm.PackageManager
import android.os.Bundle
import android.view.View
import android.view.ViewGroup
import android.widget.AdapterView
import android.widget.Spinner
import androidx.core.widget.ContentLoadingProgressBar
import androidx.core.widget.NestedScrollView
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.RecyclerView
import ru.tpu.russiantpu.R
import ru.tpu.russiantpu.dto.DocumentDTO
import ru.tpu.russiantpu.main.dataAdapters.DocumentsDataAdapter
import ru.tpu.russiantpu.utility.DownloadFileFromUrl
import ru.tpu.russiantpu.utility.SharedPreferencesService
import ru.tpu.russiantpu.utility.StartActivityService
import ru.tpu.russiantpu.utility.ToastService
import ru.tpu.russiantpu.utility.callbacks.GenericCallback
import ru.tpu.russiantpu.utility.requests.GsonService
import ru.tpu.russiantpu.utility.requests.RequestService
import java.util.*

class DocumentsFragment : Fragment(R.layout.fragment_documents) {
    private var requestService: RequestService? = null
    private var progressBar: ContentLoadingProgressBar? = null

    //переменные файла
    private var url = "" //ссылка на скачивание
    private var fileName = "" //название файла
    private var token = ""//токен для получения файла

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        // Без этой надстройки скрывается выбор списка документов
        // ("Новые документы", "просмотренные документы")
        if (view is NestedScrollView) {
            view.setFocusableInTouchMode(true)
            view.descendantFocusability = ViewGroup.FOCUS_BEFORE_DESCENDANTS;
        }

        //вспомогательные классы
        val sharedPreferencesService = SharedPreferencesService(activity)
        val gsonService = GsonService()
        val toastService = ToastService(context)
        requestService = RequestService(sharedPreferencesService, StartActivityService(activity))

        //получение информации о юзере из sharedPreferences
        val email = sharedPreferencesService.email
        val language = sharedPreferencesService.languageId
        token = sharedPreferencesService.token

        // инициализация RecyclerView с адаптером
        val docsRecyclerView: RecyclerView = view.findViewById(R.id.doc_list)
        val adapter = DocumentsDataAdapter(clickListener = { document ->
            url = document.url
            fileName = document.fileName
            DownloadFileFromUrl.downloadFile(url, fileName, token, this@DocumentsFragment)
        })
        docsRecyclerView.adapter = adapter

        val docsSelector: Spinner = view.findViewById(R.id.docs_selector)
        docsSelector.onItemSelectedListener = object : AdapterView.OnItemSelectedListener {
            override fun onItemSelected(parent: AdapterView<*>?, view: View?, position: Int, id: Long) {
                when (position) {
                    0 -> adapter.showNewDocuments()
                    1 -> adapter.showSeenDocuments()
                }
            }

            override fun onNothingSelected(parent: AdapterView<*>?) {
                // Ничего не делаем
            }
        }

        progressBar = view.findViewById(R.id.progress_bar)
        progressBar?.show() //включаем прогресс бар

        //реализация коллбека - что произойдет при получении данных с сервера
        val callback: GenericCallback<String> = object : GenericCallback<String> {
            override fun onResponse(jsonBody: String) {
                val docs = gsonService.fromJsonToArrayList(jsonBody, DocumentDTO::class.java)

                activity?.runOnUiThread {
                    progressBar?.hide() //выключаем прогресс бар
                    adapter.updateList(docs)
                }
            }

            override fun onError(message: String) {
                //выключаем прогресс бар
                activity?.runOnUiThread { progressBar?.hide() }
                toastService.showToast(message)
            }

            override fun onFailure(message: String) {
                //выключаем прогресс бар
                activity?.runOnUiThread { progressBar?.hide() }
                toastService.showToast(R.string.docs_get_error)
            }
        }

        //запрос за получение документов для пользователя (по email)
        requestService?.doRequest("document", callback, token, language, "email", email)
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        //если юзер дал разрешение на сохранение файла через диалоговое окно, пытаемся скачать файл повторно
        if (grantResults.isNotEmpty() && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
            DownloadFileFromUrl.downloadFile(url, fileName, token, this@DocumentsFragment)
        }
    }

    override fun onDetach() {
        super.onDetach()
        //при закрытии фрагмента отменяем все запросы
        requestService?.cancelAllRequests()
        //выключаем прогрессбар
        progressBar?.hide()
    }
}